invenio_pidstore/__init__.py

```diff
@@ -22,11 +22,315 @@
+persistent identifier from an external service.
+
+PIDStore comes by default with a
+:py:class:`invenio_pidstore.providers.recordid.RecordIdProvider` which will
+create Invenio legacy integer record identifiers:
+
+>>> from invenio_pidstore.providers.recordid import RecordIdProvider
+>>> provider = RecordIdProvider.create()
+>>> provider.pid.pid_type
+'recid'
+>>> provider.pid.pid_value
+'1'
+
+Creating your own provider
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+In order to create your own provider, you need to inherit from
+:py:class:`invenio_pidstore.providers.base.BaseProvider` and override the
+methods you want to customize.
+
+Here's a minimal example:
+
+>>> from invenio_pidstore.providers.base import BaseProvider
+>>> class MyProvider(BaseProvider):
+...     pid_type = 'my'
+...     pid_provider = 'test'
+...     default_status = PIDStatus.RESERVED
+
+Minters
+-------
+Minters are small functions that are responsible for minting a specific
+persistent identifier type for a specific internal object type in as automatic
+way as possible.
+
+A minter takes two arguments:
+
+- an `object uuid` of the internal object which the persistent identifier
+  should be assigned to.
+- a `dictionary object`, which the minter may extract information from, and/or
+  store extra information in. For instance, if a record object is passed, it
+  might contain the DOI which a provider should mint, or alternatively, if the
+  persistent identifier is generated, the minter can store the generated value
+  into the record.
+
+Below is an example minter which mints external UUIDs (not to be confused with
+the internal object UUIDs), and stores the generated UUID into the dictionary
+object (which could be e.g. a record):
+
+>>> import uuid
+>>> def my_uuid_minter(object_uuid, data):
+...     pid = PersistentIdentifier.create('myuuid', str(uuid.uuid4()))
+...     data['uuid'] = str(pid.object_uuid)
+...     return pid
+
+Registering minters
+~~~~~~~~~~~~~~~~~~~
+Minters are usually used by other modules and thus registered on the
+Flask application.
+
+First import the proxy to the current application's PIDStore:
+
+>>> from invenio_pidstore import current_pidstore
+
+Next, register a minter:
+
+>>> current_pidstore.register_minter('my_uuid_minter', my_uuid_minter)
+
+You can retrieve registered minters from the extension as well:
+
+>>> current_pidstore.minters['my_uuid_minter']
+<function my_uuid_minter at ...>
+
+Entry points loading
+~~~~~~~~~~~~~~~~~~~~
+PIDStore will automatically register minters defined by the entry point group
+``invenio_pidstore.minters``.
+
+Example:
+
+.. code-block:: python
+
+   # setup.py
+   setup(
+       # ...
+       entry_points={
+           'invenio_pidstore.minters': [
+               'recid_minter = invenio_pidstore.minters:recid_minter',
+           ]})
+
+Above is equivalent to:
+
+.. code-block:: python
+
+   from invenio_pidstore.minters import recid_minter
+   current_pidstore.register_minter('recid_minter', recid_minter)
+
+"""
```

Comment:
@lnielsen thanks a lot for this piece of documentation. It is really informative! :clap: